### ISSUE TITLE
fail to connect any irc server with irssi running minoca instance - [Interrupted system call]

### ISSUE THREAD

**Issue by melezhik** [minoca/os#72]
Hi!

I have successfully built an irssi package and run it on minoca OS instance:

```
bash-4.3# irssi --version
irssi 0.8.20 (20160914 1355)

```

But when I am trying to connect it to some IRC server :

```
irssi -c chat.freenode.net -p 6667
```

I get this:

```
Irssi v0.8.20 - http://www.irssi.org                                                                                                                    
03:54 -!- Irssi: Looking up chat.freenode.net
03:54 -!- Irssi: Connecting to chat.freenode.net [91.217.189.42] port 6667
03:54 -!- Irssi: Unable to connect server chat.freenode.net port 6667 [Interrupted system call]

```

The same error with any other irc servers ( irc.oftc.net , so on )

My network looks fine:

```

bash-4.3# curl -L -D - http://minocacorp.com/ -o /dev/null
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0HTTP/1.1 301 Moved Permanently
Date: Thu, 01 Dec 2016 11:53:17 GMT
Server: Apache/2.4.7 (Ubuntu)
Location: http://www.minocacorp.com/
Content-Length: 313
Content-Type: text/html; charset=iso-8859-1

100   313  100   313    0     0    753      0 --:--:-- --:--:-- --:--:--  1719
HTTP/1.1 200 OK
Date: Thu, 01 Dec 2016 11:53:18 GMT
Server: Apache/2.4.7 (Ubuntu)
Vary: Cookie,Accept-Encoding
X-Frame-Options: SAMEORIGIN
Transfer-Encoding: chunked
Content-Type: text/html; charset=utf-8

100 10501    0 10501    0     0  12723      0 --:--:-- --:--:-- --:--:-- 12723
bash-4.3# curl chat.freenode.net:6667                     
:asimov.freenode.net NOTICE * :*** Looking up your hostname...
:asimov.freenode.net NOTICE * :*** Checking Ident
:asimov.freenode.net NOTICE * :*** Found your hostname
:asimov.freenode.net NOTICE * :*** No Ident response
ERROR :Closing Link: 127.0.0.1 (Client Exit)

```

**Comment by melezhik** [comment]
Adding @vascocosta here

**Comment by vascocosta** [comment]
I still cannot reproduce this bug, it connects without issues here. However, I'm using qemu on Windows, which seems to be the only difference between our environment.

It would be interesting to run the Minoca kernel debugger for this one, as it seems to be related to some networking system call. It only runs on Windows though, from what I know, so I can't really debug the kernel when this issue happens.

Nevertheless I'll try to reproduce it by running qemu on Linux.

**Comment by melezhik** [comment]
another strange thing, somehow, only ONCE I succeeded to connect irssi client to Freenode , but then kept going with the same errors - Interrupted system call when doing next attempts, looks like flickering bug ...

**Comment by ccstevens** [comment]
I'm going to take a look at this after I get to the bottom of issue #62.

@melezhik Are you running on QEMU or real hardware?

@vascocosta If you've got QEMU on Linux running inside a VM, you can actually add a serial port with a named pipe (e.g. `\\.\pipe\vmware_minoca`) and debug from Windows. Setup includes:

1. Run the following from the Linux command line assuming the serial device you added is `/dev/ttyS1`:

       sudo stty -F /dev/ttyS1 -ocrnl -onlcr -onlret

2. Add the `-serial /dev/ttyS1` to your Linux QEMU command.
3. Update your Minoca Debugger `-k` option to point to the named pipe.

       run_in_background $SRCROOT/x86$DEBUG/tools/bin/debugui.exe -s $SRCROOT/x86$DEBUG/bin -e $SRCROOT/x86$DEBUG/tools/bin/kexts.dll -k '\\.\pipe\vmware_minoca'

And, you're right. The Minoca debugger doesn't run on Linux.

**Comment by melezhik** [comment]
Hi Evan!

> @melezhik Are you running on QEMU or real hardware?

I am running on qemu ... with host system under Linux ( Ubuntu )

**Comment by vascocosta** [comment]
I managed to set up a pipe as per @ccstevens instructions, therefore connecting the kernel debugger to a qemu instance running inside a Linux VirtualBox VM.

Here's some debugging I did:

```
kd>bp NetpTcpConnect
kd>g
Breakpoint 0 hit!
netcore!NetpTcpConnect:
80ad6696 push	ebp	; 55
kd>kn
No Frame    RetAddr  Call Site
 0 80d8fd70 80ad25d9 netcore!NetpTcpConnect
 1 80d8fda0 800651cd netcore!NetConnect+0x5c
 2 80d8fdd0 8006580b kernel!IoSocketConnect+0x63
 3 80d8fe30 8003a8e8 kernel!IoSysSocketConnect+0x76
 4 80d8ffb8 7fffea40 kernel!KeSystemCallHandler+0x10e
 5 80d8ff78 8009295f 0x7fffea40
 6 7fffeab8 773d2d3a kernel+0x9195f
 7 7fffeb38 080fd812 0x773d2d3a
 8 7fffeb98 081019a4 0x080fd812
 9 7fffec08 08101ca6 0x081019a4
10 7fffec78 080f96e3 0x08101ca6
11 7fffeca8 7762bc40 0x080f96e3
12 7fffecd8 775dc6ef 0x7762bc40
13 7fffed28 775dd48d 0x775dc6ef
14 7fffed48 775dd672 0x775dd48d
15 7fffed88 775dd73a 0x775dd672
16 7fffedb8 0809b35f 0x775dd73a
17 7fffede8 0822a09e 0x0809b35f
18 7fffee24 777d6f6d 0x0822a09e
19 7fffee74 00000000 0x777d6f6d
kd>dv
[0x80d8fd70] Socket              : 0x80883778
[0x80d8fd74] Address             : 0x80d8fe84

[0x80d8fd4c] ReturnedEvents      : 0x80ad6696
kd>dt ReturnedEvents
0x80ad6696
```

Does the value of `ReturnedEvents` tell you anything meaningful, @ccstevens?

**Comment by ccstevens** [comment]
@vascocosta , `ReturnedEvents` is a local and not yet initialized. The break point you set halts the system at the top of the `NetpTcpConnect` function and from your output, it doesn't look like you've used `p` or `t` to step through any instructions. As a result, `ReturnedEvents` is just storing garbage (looks like a kernel address). The expected values in `ReturnedEvents` are the [`POLL_EVENT_*`](https://github.com/minoca/os/blob/5f082d58fabc47c459eea05968987c4c65c848ff/include/minoca/kernel/syscall.h#L232) flags, but won't be filled in until after `IoWaitForIoObjectState` is called.

You could run `gu` (go up one stack frame) and see what `NetpTcpConnect` returns by running the `r eax` command. That will be a [status code](https://github.com/minoca/os/blob/5f082d58fabc47c459eea05968987c4c65c848ff/include/minoca/lib/status.h). I'd imagine, to match @melezhik's output, it would be `STATUS_INTERRUPTED` (-4, a.k.a. 0xFFFFFFFC). 

Now, to figure out what caused the interruption we'll have to look at the process's pending signals. The debug steps to find the pending signals are below. You should be in `NetConnect` at this point.

1. `proc` - dumps the current processor information structure. Scroll up and find the `RunningThread` member at offset `0x14`.
2. `dt _kthread 0xXXXXXXXX` - where `0xXXXXXXXX` is the address you gathered from `RunningThread`.
3.  Find the `PendingSignals` at offset `0x78`. Are there any bits set? What about in `BlockedSignals` at offset `0x80`?
4. `dt _kprocess 0xYYYYYYYY` - where `0xYYYYYYYY` is the address taken from the thread's `OwningProcess` member (offset `0x40`).
5. Are there any pending signals listed under `PendingSignals` at offset `0xd8`? Any `IgnoredSignals` at offset `0xe0`?

The `PendingSignals` fields are bitmasks, so you can determine which signal interrupted by looking at how [`ADD_SIGNAL`](https://github.com/minoca/os/blob/5f082d58fabc47c459eea05968987c4c65c848ff/include/minoca/kernel/ksignals.h#L40) works and checking against the [`SIGNAL_*`](https://github.com/minoca/os/blob/5f082d58fabc47c459eea05968987c4c65c848ff/include/minoca/kernel/ksignals.h#L141) values.

If you follow that, you're well on your way to becoming a debug expert. But let me know if any of it doesn't make sense.

**Comment by vascocosta** [comment]
@ccstevens,

Thanks for the detailed tutorial, it was extremely helpful for this particular case and enlightening within the greater picture of general kernel debugging.

So, first off, I went up one stack frame and confirmed that `NetpTcpConnect` is returning `0xFFFFFFFC` as expected.

Next, inside `NetConnect`, I followed your steps and verified that the `RunningThread`'s `PendingSignals` and `BlockedSignals` were both unset (0x0). Moving on to the thread's `OwningProcess` member I finally found something interesting:

```
  +0xd8   PendingSignals    : 0x8000
  +0xe0   IgnoredSignals    : 0x1010
```

From what I can tell, `ADD_SIGNAL` sets `_SignalSet` with the result of a bitwise or between the previous `_SignalSet` value and `(1 << (_SignalNumber) - 1)`. It therefore follows that if I know two `_SignalSet` values before and after some `_SignalNumber` is added, I can find which signal it was. I'm assuming I can use the current value of `PendingSignals` inside `NetConnect` as the previous `_SignalSet` value and then see what the `_SignalSet` becomes once `NetpTcpConnect` is called and use it as the next `_SignalSet` value. Did I get this right or am I complicating it more than I should?

**Comment by ccstevens** [comment]
@vascocosta 

Awesome! You've got the right thought. But since `PendingSignals` only has one bit set, you can safely assume it was `0x0` before the call to `NetpTcpConnect`. And that bit `0x8000` isn't ignored, which means the thread/process got woken up from its wait for that signal. So the remaining puzzle is to figure out which signal `0x8000` corresponds to.

**Comment by vascocosta** [comment]
If I got it right:

`#define SIGNAL_CHILD_PROCESS_ACTIVITY 16`

**Comment by ccstevens** [comment]
@melezhik stay tuned... @vascocosta and I discussed this over IRC and he's hot on the trail of this bug!

**Comment by melezhik** [comment]
Ok, good to hear, guys !

**Comment by vascocosta** [comment]
@ccstevens,

Irssi implements a nonblocking `gethostbyname` of sorts by forking a worker child that does the lookup. The interrupted system call happens when the child exits with `_exit(99)`. I did a simple test to verify my debugging results by trying to connect to the IRC server using the IP address, instead of the hostname, therefore bypassing irssi's nonblocking lookup code. As expected, `NetpTcpConnect` never gets interrupted this way.

Irssi doesn't call `sigaction` with `SIGCHLD` anywhere, nevertheless, from what I've read on the manual page, `_exit` does send a `SIGCHLD` to the caller's parent, thus interrupting Minoca's kernel.

What would be the optimal course of action here? Using `exit` instead?

Here's the source code for irssi's nonblocking `gethostbyname`:

```c
/* nonblocking gethostbyname(), ip (IPADDR) + error (int, 0 = not error) is
   written to pipe when found PID of the resolver child is returned */
int net_gethostbyname_nonblock(const char *addr, GIOChannel *pipe,
                               int reverse_lookup)
{
        RESOLVED_IP_REC rec;
        const char *errorstr;
        int pid;
        int len;

        g_return_val_if_fail(addr != NULL, FALSE);

        pid = fork();
        if (pid > 0) {
                /* parent */
                pidwait_add(pid);
                return pid;
        }

        if (pid != 0) {
                /* failed! */
                g_warning("net_connect_thread(): fork() failed! "
                          "Using blocking resolving");
        }

        /* child */
        srand(time(NULL));

        memset(&rec, 0, sizeof(rec));
        rec.error = net_gethostbyname(addr, &rec.ip4, &rec.ip6);
        if (rec.error == 0) {
                errorstr = NULL;
                if (reverse_lookup) {
                        /* reverse lookup the IP, ignore any error */
                        if (rec.ip4.family != 0)
                                net_gethostbyaddr(&rec.ip4, &rec.host4);
                        if (rec.ip6.family != 0)
                                net_gethostbyaddr(&rec.ip6, &rec.host6);
                }
        } else {
                errorstr = net_gethosterror(rec.error);
                rec.errlen = errorstr == NULL ? 0 : strlen(errorstr)+1;
        }

        g_io_channel_write_block(pipe, &rec, sizeof(rec));
        if (rec.errlen != 0)
                g_io_channel_write_block(pipe, (void *) errorstr, rec.errlen);
        else {
                if (rec.host4) {
                        len = strlen(rec.host4) + 1;
                        g_io_channel_write_block(pipe, (void *) &len,
                                                       sizeof(int));
                        g_io_channel_write_block(pipe, (void *) rec.host4,
                                                       len);
                }
                if (rec.host6) {
                        len = strlen(rec.host6) + 1;
                        g_io_channel_write_block(pipe, (void *) &len,
                                                       sizeof(int));
                        g_io_channel_write_block(pipe, (void *) rec.host6,
                                                       len);
                }
        }

        if (pid == 0)
                _exit(99);

        /* we used blocking lookup */
        return 0;
}
```

**Comment by ccstevens** [comment]
@vascocosta,

Looks like the right block of code to explain what we are seeing! Switching `_exit(99)` to `exit(99)` won't actually change this behavior as they filter down to the same system call, which sends a `SIGCHLD` to the parent process. But, by default, `SIGCHLD` is supposed to be ignored! So the question remains to figure out what is calling `sigaction` with `SIGCHLD`. If not irssi, then what? There has to be something, as we saw in the debugger that the `HandledSignals` field had the `SIGCHLD` bit set. My hunch? Investigate how `pidwait_add` works...

**Comment by evangreen** [comment]
I'm jumping in right in the middle here with very little context, so I may be way off base here, but maybe check to see if SIGCHLD is marked as handled when irssi first starts running. Maybe the shell failed to reset that signal disposition? Or maybe the kernel failed to reset it? I forget if signal dispositions are supposed to be reset on exec().

But if you see that signal marked as "handled" before irssi even gets to run code, then we know it was launched incorrectly and should look at whoever launched it.

**Comment by vascocosta** [comment]
After some time I also thought that could be the case @evangreen, but I did an strace on Linux and it was clear that irssi marks `SIGCHLD` as handled itself.

Here's the trace:

```
rt_sigaction(SIGCHLD, {0x7ff66d5ed370, [], SA_RESTORER|SA_RESTART|SA_NOCLDSTOP, 0x7ff66dd008d0}, NULL, 8) = 0
wait4(1111, 0x2710984, WNOHANG, NULL)   = 0
write(3, "\1\0\0\0\0\0\0\0", 8)         = 8
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3439, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3439, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3439, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3439, ...}) = 0
write(3, "\1\0\0\0\0\0\0\0", 8)         = 8
write(1, "\33[?25l\33[?1c\33[22;1H00:23 \33[34m\33[1"..., 177) = 177
write(1, "\33[37m\33[44m\33[?25l\33[?1c\33[24;9H    "..., 226) = 226
poll([{fd=0, events=POLLIN|POLLPRI}, {fd=3, events=POLLIN}, {fd=4, events=POLLIN|POLLPRI}], 3, 156) = 1 ([{fd=3, revents=POLLIN}])
poll([{fd=0, events=POLLIN|POLLPRI}, {fd=3, events=POLLIN}, {fd=4, events=POLLIN|POLLPRI}], 3, 156) = 1 ([{fd=3, revents=POLLIN}])
read(3, "\3\0\0\0\0\0\0\0", 16)         = 8
poll([{fd=0, events=POLLIN|POLLPRI}, {fd=3, events=POLLIN}, {fd=4, events=POLLIN|POLLPRI}], 3, 156) = 1 ([{fd=4, revents=POLLIN}])
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=1111, si_uid=1000, si_status=99, si_utime=0, si_stime=0} ---
write(6, "\1\0\0\0\0\0\0\0", 8)         = 8
rt_sigreturn()                          = 1
read(3, "\1\0\0\0\0\0\0\0", 16)         = 8
write(3, "\1\0\0\0\0\0\0\0", 8)         = 8
fcntl(4, F_SETFL, O_RDONLY|O_NONBLOCK)  = 0
read(4, "\2\0\0\0mJ\310]\0\0\0\0\0\0\0\0\0\0\0\0\n\0\0\0 \1\6\260\0\16*\30"..., 72) = 72
close(4)                                = 0
close(5)                                = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3439, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3439, ...}) = 0
socket(PF_INET, SOCK_STREAM, IPPROTO_IP) = 4
fcntl(4, F_SETFL, O_RDONLY|O_NONBLOCK)  = 0
setsockopt(4, SOL_SOCKET, SO_REUSEADDR, [1], 4) = 0
setsockopt(4, SOL_SOCKET, SO_KEEPALIVE, [1], 4) = 0
connect(4, {sa_family=AF_INET, sin_port=htons(6667), sin_addr=inet_addr("109.74.200.93")}, 16) = -1 EINPROGRESS (Operation now in progress)
fstat(4, {st_mode=S_IFSOCK|0777, st_size=0, ...}) = 0
fcntl(4, F_GETFL)                       = 0x802 (flags O_RDWR|O_NONBLOCK)
write(3, "\1\0\0\0\0\0\0\0", 8)         = 8
write(1, "\33[?25l\33[?1c\33[23;1H00:23 \33[34m\33[1"..., 276) = 276
poll([{fd=0, events=POLLIN|POLLPRI}, {fd=3, events=POLLIN}, {fd=4, events=POLLIN|POLLPRI|POLLOUT}], 3, 0) = 1 ([{fd=3, revents=POLLIN}])
rt_sigaction(SIGCHLD, {SIG_DFL, [], SA_RESTORER, 0x7ff66dd008d0}, NULL, 8) = 0
poll([{fd=0, events=POLLIN|POLLPRI}, {fd=3, events=POLLIN}, {fd=4, events=POLLIN|POLLPRI|POLLOUT}], 3, 113) = 1 ([{fd=3, revents=POLLIN}])
read(3, "\2\0\0\0\0\0\0\0", 16)         = 8
poll([{fd=0, events=POLLIN|POLLPRI}, {fd=3, events=POLLIN}, {fd=4, events=POLLIN|POLLPRI|POLLOUT}], 3, 113) = 1 ([{fd=4, revents=POLLOUT}])
read(3, 0x7ffdcaa41370, 16)             = -1 EAGAIN (Resource temporarily unavailable)
write(3, "\1\0\0\0\0\0\0\0", 8)         = 8
getsockopt(4, SOL_SOCKET, SO_ERROR, [0], [4]) = 0
```

**Comment by vascocosta** [comment]
And the culprit is... **Glib**!

Glib handles `SIGCHLD` within some of its functions. Just like @ccstevens suggested, the prime suspect was indeed `pidwait_add`. It basically calls some glib utility functions to watch the forked child. Commenting `pidwait_add` out seems to work and I see no more `sigaction(SIGCHLD)` in strace.

I do wonder, however, why on Linux this isn't really an issue and why I've never seen an "Interrupted system call" error on Linux after using irssi for 17 years there. What's fundamentally different between Minoca OS and Linux that allows the system call to presumably be resumed on Linux?

```c
/* add a pid to wait list */
void pidwait_add(int pid)
{
        if (g_hash_table_lookup(child_pids, GINT_TO_POINTER(pid)) == NULL) {
                int id = g_child_watch_add_full(10, pid, sig_child, NULL, NULL);
                g_hash_table_insert(child_pids, GINT_TO_POINTER(pid), GINT_TO_POINTER(id));
                pids = g_slist_append(pids, GINT_TO_POINTER(pid));
        }
}

static void sig_child(GPid pid, gint status, gpointer data)
{
        signal_emit_id(signal_pidwait, 2, GINT_TO_POINTER(pid),
                       GINT_TO_POINTER(status));
        g_hash_table_remove(child_pids, GINT_TO_POINTER(pid));
        pids = g_slist_remove(pids, GINT_TO_POINTER(pid));
}
```

**Comment by melezhik** [comment]
HI @vascocosta , just want to be sure, are you able to reproduce this issue on your side?

**Comment by vascocosta** [comment]
Yes @melezhik, I've been reproducing it the whole weekend on Linux. We're close to a definitive fix.

**Comment by melezhik** [comment]
ah, ok ))) I just wanted to be sure that the issue is not specific for the build I made for irss, sorry for "silly" question,  I believe you will find the clue soon!

**Comment by vascocosta** [comment]
@melezhik,

By saying it wasn't really an issue on Linux on my previous post what I meant was running irssi directly on Linux, instead of using Minoca on a qemu instance on Linux. I mentioned that because I used irssi on my Linux desktop since 1999. :)

**Comment by melezhik** [comment]
yeah, ok @vascocosta  ))))

**Comment by ccstevens** [comment]
@vascocosta good sleuthing.

> What's fundamentally different between Minoca OS and Linux that allows the system call to presumably be resumed on Linux?

A bug.

The Linux trace was a big hint for this one. Glib must be calling `sigaction` with `SA_RESTART` set. Networking actions like `accept`, `connect`, `recv` and `send` are supposed to be restarted when `SA_RESTART` is set. Alas, it looks like something is missing in [`IoSysSocketConnect`](https://github.com/minoca/os/blob/5f082d58fabc47c459eea05968987c4c65c848ff/kernel/io/socket.c#L1488) that exists in [`IoSysSocketAccept`](https://github.com/minoca/os/blob/5f082d58fabc47c459eea05968987c4c65c848ff/kernel/io/socket.c#L1352). 

Do you see it @vascocosta?

**Comment by vascocosta** [comment]
@ccstevens,

Assuming `IopConvertInterruptedSocketStatus` checks whether a system call can be restarted before resetting the status, would it be this?

```c
    if (Status == STATUS_INTERRUPTED) {
        Status = IopConvertInterruptedSocketStatus(IoHandle, 0, FALSE);
    }
```

**Comment by ccstevens** [comment]
Yup, that's what it does. If you look at the implementation of IopConvertInterruptedSocketStatus, you'll want to be careful with the third argument. SA_RESTART isn't valid on "input socket" calls (accept or recv) if a receive timeout is set (via setsockopt) and it's not valid on "output socket" calls (connect or send) if a send timeout is set.

But I think that should fix the issue!

On Dec 4, 2016, at 10:12, Vasco Costa <notifications@github.com <mailto:notifications@github.com> > wrote:

@ccstevens,

Assuming IopConvertInterruptedSocketStatus checks whether a system call can be restarted before resetting the status, would it be this?


    if (Status == STATUS_INTERRUPTED) {
        Status = IopConvertInterruptedSocketStatus(IoHandle, 0, FALSE);
    }

**Comment by vascocosta** [comment]
@melezhik,

I will then try the kernel fix as proposed by @ccstevens on a local branch, test everything thoroughly and if the issue is gone and nothing else breaks, I'll send a pull request.

**Comment by melezhik** [comment]
@vascocosta sounds good

**Comment by vascocosta** [comment]
@ccstevens,

Setting `Status = IopConvertInterruptedSocketStatus(IoHandle, 0, TRUE);` like you can see below was apparently working under all circumstances on qemu on Windows, however on Linux I'm now getting a different error:

`connection already in progress` which is code `EALREADY`, instead of `the system call was interrupted`. Any ideas on what could be going on?

```c
SysSocketConnectEnd:
    if (PathCopy != NULL) {
        MmFreePagedPool(PathCopy);
    }

    //
    // An interrupted socket connect cannot be restarted if a send timeout
    // has been set.
    //
```
```c
    if (Status == STATUS_INTERRUPTED) {
        //Status = STATUS_RESTART_NO_SIGNAL;
        Status = IopConvertInterruptedSocketStatus(IoHandle, 0, TRUE);
    }
```
```c
    //
    // Release the reference that was added when the handle was looked up.
    //

    if (IoHandle != NULL) {
        IoIoHandleReleaseReference(IoHandle);
    }

    return Status;
}
```

**Comment by ccstevens** [comment]
If I had to guess, `EALREADY` is coming from `NetpTcpConnect` returning `STATUS_ALREADY_INITIALIZED` ([here](https://github.com/minoca/os/blob/83647ce4c25068aba0bc53be66c30234458e2408/drivers/net/netcore/tcp.c#L1286)). 

If you study what's going on in that routine, it doesn't look like things will go well if it gets called twice with the same socket. And that's exactly what an interrupt + restart would do. The first time it gets called, the socket will transition from `TcpStateInitialized` to `TcpStateSynSent` and then wait for the connection to be established. If that wait gets interrupted as we were seeing, the socket will already be in the `TcpStateSynSent` state when `NetpTcpConnect` gets called on restart. So, it will fail at the line above.

So, I suppose we need to make `NetpTcpConnect` work when it gets called a second time. Once we've sent the `SYN` packet and are in the `TcpStateSynSent` state, the background network processing can move the socket to one of the following states: `TcpStateSynReceived`, `TcpStateEstablished` or `TcpStateCloseWait` states. We need to account for all of those states on the restart. I guess it's OK to call `NetpTcpConnect` if a socket is already in one of those states - you wouldn't call `NetpTcpSetState` and just skip to the wait that got interrupted. 

But, as `connect()` is supposed to return `EALREADY` if a connect request is already in progress, we may want to know the difference between a restart and a second call to connect. Or only allow one extra call to succeed. The `TCP_SOCKET.Flags` field may be of use here. We could mark that an interrupt occurred and only allow a second connect call if a certain flag is set. But that still allows a race between the thread that is restarting the system call and some other thread trying to call connect. I guess if someone builds an application with that race, they would need to be able to handle either instance of `connect()` returning `EALREADY`. 

Is this something you want to take a stab at fixing too?

**Comment by ccstevens** [comment]
@vascocosta My TCP connect fix is in [7fe29781](https://github.com/minoca/os/commit/7fe29781d9676175a13d80aa5c4700bce1b7acad) in case you want to take a look. 

@melezhik Feel free to sync to the latest source and give irssi a try!

**Comment by melezhik** [comment]
@ccstevens  sure, will try soon.

**Comment by melezhik** [comment]
@ccstevens at brief look the code (I am not a C/C++ programmer but ... ) sound like it was en essential bug? huh?

**Comment by vascocosta** [comment]
@ccstevens,

I got the first part right, up to line 1309. I also figured the interrupt flag should be cleared before trying connect once again. After that, my current knowledge of the TCP state machine wasn't enough, thus it was a good call to hand it over to you right away.

Having said that, this reignited my interest in networking, so I'll be revisiting TCP/IP soon.

And last but not the least, good job!

**Comment by ccstevens** [comment]
@vascocosta If you had the first part right and thought to clear the flag, you weren't too far off. The key was to not call `NetpTcpSetState` with `TcpStateSynSent` again, as that would mess up the state machine and hit an assert.

Feel free to close this one out if `irssi` is working (or at least this issue).

**Comment by vascocosta** [comment]
I'm currently unable to reproduce this issue to test the fix due to issue #86. Hopefully @melezhik can test it and close it if the error is gone.

**Comment by melezhik** [comment]
Hi @evangreen @ccstevens ! I am trying to rebuild irssi from the scrathc but get this ( on make image stage):

```
rm -f /home/melezhik/minoca/x86dbg/bin/install.img /home/melezhik/minoca/x86dbg/bin/pc.img /home/melezhik/minoca/x86dbg/bin/pcefi.img /home/melezhik/minoca/x86dbg/bin/pctiny.img 
rm -rf /home/melezhik/minoca/x86dbg/obj/os/images
Building - install.img
/bin/sh: 1: msetup: not found
make[1]: *** [/home/melezhik/minoca/x86dbg/bin/install.img] Error 127
make: *** [/home/melezhik/minoca/x86dbg/obj/os/images] Error 2
```

I have not seen this before and only get after getting new update for os repository... latest commit - 6243af9d0656d9e4d2d2ef9c633bf283b85e9303 )

**Comment by evangreen** [comment]
Is msetup built? You should see it at `/home/melezhik/minoca/x86dbg/tools/bin/msetup`. Is `/home/melezhik/x86dbg/tools/bin` in your `PATH`?

**Comment by melezhik** [comment]
Hi Evan! Every time I run my minoca-build plugin I get different errors, so unlike I am :)))) , let me start all over the scratch ( make wipe, make clean so on ) and get back here  ...

**Comment by melezhik** [comment]
Hmm, this is error I get trying to create a fresh minoca build ( after gettings updates from github/gitlab )

https://gist.github.com/melezhik/f8e45a01a6c91ba0982873f9d0c334b5

**Comment by melezhik** [comment]
Sorry if remind you but I am trying to build on Linux:

```
Ubuntu 14.04.5 LTS \n \l
```

**Comment by evangreen** [comment]
I just pushed a fix for the fgets compile error. Can you pull the latest and try again?

**Comment by vascocosta** [comment]
@ccstevens and @melezhik,

I'm running irssi without any issues after the TCP fix on both the latest version of QEMU on Linux and most importantly, for the first time ever, [natively on my Raspberry Pi 3](http://i.imgur.com/qvx8coZ.png)! The "interrupted system call" error is gone for good.

Feel free to close this issue, @melezhik.

**Comment by melezhik** [comment]
Ok, Evan , I think we can close this issue, after @ccstevens last comment. I will check your fix later , anyway it is irrelevant to this task and if I get a troubles I will start new issue.

**Comment by melezhik** [comment]
Hi @vascocosta @ccstevens @evangreen  I have successfully built irssi on ubuntu and run it without any errors on qemu minoca instance , now I am able to run /connect command without any errors. Confirming issue is gone from my point of view.

**Comment by vascocosta** [comment]
It's gone here too.

**Comment by ccstevens** [comment]
Cool. Thanks for following up on this!

**Comment by diveyez** [comment]
Want to see it in action, use it with torscks, Is there a PERMANENT fix for this yet?

**Comment by ccstevens** [comment]
Hey @diveyez! Commit [7fe29781](https://github.com/minoca/os/commit/7fe29781d9676175a13d80aa5c4700bce1b7acad) was meant to be a permanent fix for this issue. Are you stil seeing a problem?
